src/Fixer/Import/FullyQualifiedStrictTypesFixer.php

```diff
@@ -280,7 +330,150 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
     /**
      * @param array<string, string> $uses
      */
-    private function fixFunction(FunctionsAnalyzer $functionsAnalyzer, Tokens $tokens, int $index, array &$uses, string $namespaceName): void
+    private function refreshUsesCache(array $uses): void
```

Comment:
The cache is great, as we have two phases, in one we collect all uses, in next, we fill the cache and use it massively.

Reply:
I know there must be 2 iterations, I just think it could be done differently, without refreshing the cache so many times during processing. I would see it rather like:

- determine all the namespace blocks
- for each namespaced block:
  - collect current imports (respecting import kinds)
  - collect defined symbols and implicit imports (relative FQNs), also respecting kinds
  - for each found symbol:
    - register it for importing (if needed), this should just update collected data like it's done now. This most probably would improve the process comparing to your solution, because you would only add single items to existing lists/collections, instead of iterating through whole `$uses` every time.
    - shorten symbols where possible (analyse found FQN against collected information)
  - import newly registered symbols

This way you don't need to refresh cache multiple times, you build all the lists once during collecting phase, and then you only do the shortening based on collected data. This is something I [suggested early](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7679#discussion_r1442644000) 😉.

Reply:
The cache is refreshed only if it needs to - https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/7aeb112384d3fecc1f5fe55f822eecbbba23034e/src/Fixer/Import/FullyQualifiedStrictTypesFixer.php#L335 ;-)

Reply:
I'll verify again, because I don't remember if I tested inside or outside the `if`. Please keep discussion open until verified.

Reply:
fixed, the type was correctly resolved but the replacing code was broken - https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/commit/0ee915aa4

Reply:
OK, for `import new symbols from all supported places` cache is actually refreshed 4 times, so 👍.